### PR TITLE
Strings Support

### PR DESCRIPTION
**Pull request by EduardoLR10**

### CODE REVIEW

File: Chime/Parser.m [hunk 1]
```diff
@@ -233,11 +322,22 @@ - (NSMutableArray *)passTwo {
...
 
+NSArray *cleanProgram(NSString *stringProgram) {
+  NSMutableArray *trimmedProgram = [NSMutableArray array];
+  for (NSString *line in [stringProgram componentsSeparatedByString:@"\n"]) {
+    NSString *trimmedLine =
+        [line stringByTrimmingCharactersInSet:[NSCharacterSet
+                                                  whitespaceCharacterSet]];
+    [trimmedProgram addObject:trimmedLine];
+  }
+  return [trimmedProgram copy];
```
Comment (z-silver) [inline Chime/Parser.m:339]: Is there any particular reason we are copying the array instead of just returning it?

File: Chime/Parser.m [hunk 2]
```diff
@@ -193,9 +238,51 @@ - (void)handleInstructions:(BOOL)isLabel
...
+  NSMutableArray *stringPayload = packString(content);
+  if (length > MAX_STRING_SIZE) {
+    @throw [NSException
+        exceptionWithName:@"String limit exceeded"
+                   reason:[NSString
+                              stringWithFormat:
+                                  @"String \"%@\" exceeded max size of %d",
+                                  content, MAX_STRING_SIZE]
                  userInfo:nil];
   }
```
Comment (z-silver) [inline Chime/Parser.m]: I would swap this local variable and this `if`.

File: Chime/Parser.m [hunk 3]
```diff
@@ -193,9 +238,51 @@ - (void)handleInstructions:(BOOL)isLabel
...
+  [bytecodes addObject:@((uint32_t)length)];
+  [bytecodes addObjectsFromArray:[stringPayload copy]];
```
Comment (z-silver) [inline Chime/Parser.m:286]: Any particular reason to copy the payload here?

File: Chime/Parser.m [hunk 4]
```diff
@@ -193,9 +238,51 @@ - (void)handleInstructions:(BOOL)isLabel
...
+                                        @(_physicalLinesCounter), string]
+                 userInfo:nil];
+  }
+}
+
+NSMutableArray *packString(NSString *string) {
+  NSMutableArray *pack = [[NSMutableArray alloc] init];
+  NSUInteger counter = 0;
+  uint32_t buffer = 0;
+  NSUInteger wordSize = 4;
```
Comment (z-silver) [inline Chime/Parser.m]: This should be defined in a wider scope.

File: Chime/Parser.m [hunk 5]
```diff
@@ -109,22 +119,57 @@ - (void)handleVariableLine:(NSString *)line {
...
+        exceptionWithName:@"Invalid String"
+                   reason:[NSString
+                              stringWithFormat:
+                                  @"Found string unbalanced delimiter at %lu",
+                                  _physicalLinesCounter]
+                 userInfo:nil];
+  }
+
+  NSUInteger length = line.length;
+  NSUInteger wordSize = 4;
```
Comment (z-silver) [inline Chime/Parser.m]: This should be defined in a wider scope.

File: Chime/Parser.m [hunk 6]
```diff
@@ -109,22 +119,57 @@ - (void)handleVariableLine:(NSString *)line {
...
 }
 
+BOOL isChimeString(NSString *string) {
+  return isStringStartingWith(string, @"\"");
+}
+
+BOOL checkBothStringEnds(NSString *string) {
+  return ('\"' == [string characterAtIndex:0]) &&
+         ('\"' == [string characterAtIndex:(string.length - 1)]);
+}
```
Comment (z-silver) [inline Chime/Parser.m:130]: Small observation here: when we get around to adding escape codes, we can check the end of the string by insisting that the last `"` character be preceded by an even number of `\`.

**Review by z-silver — CHANGES_REQUESTED** [review]

File: Chime/Parser.m (hunk 1 above)
Comment (EduardoLR10) [reply Chime/Parser.m:339]: I believe this has something to do with converting `NSMutableArray` to `NSArray`.

File: Chime/Parser.m (hunk 3 above)
Comment (EduardoLR10) [reply Chime/Parser.m:286]: I believe this has something to do with converting `NSMutableArray` to `NSArray`.

**Review by z-silver — APPROVED** [review]

**Pull request merged**